contrib/palantirjavaformat/src/mill/contrib/palantirjavaformat/PalantirJavaFormatModule.scala

```diff
@@ -0,0 +1,187 @@
+package mill
+package contrib.palantirjavaformat
+
+import mill.api.Loose
+import mill.api.PathRef
+import mill.define.{Discover, ExternalModule}
+import mill.main.Tasks
+import mill.scalalib.DepSyntax
+import mill.scalalib.JavaModule
+import mill.util.Jvm
+
+/**
+ * Formats Java source files using [[https://github.com/palantir/palantir-java-format Palantir Java Format]].
+ */
+trait PalantirJavaFormatModule extends JavaModule {
+
+  /**
+   * Formats Java source files.
+   *
+   * @param check if an exception should be raised when formatting errors are found
+   *              - when `true`, files are not formatted
+   * @param sources list of file or folder path(s) to be processed
+   *                - path must be relative to [[millSourcePath]]
+   *                - when empty, all [[sources]] are processed
+   */
+  def javafmt(
+      @mainargs.arg check: Boolean = false,
+      sources: mainargs.Leftover[String]
+  ): Command[Unit] = Task.Command {
+
+    val _sources =
+      if (sources.value.isEmpty) this.sources()
+      else sources.value.iterator.map(rel => PathRef(millSourcePath / os.RelPath(rel)))
+
+    PalantirJavaFormatModule.palantirAction(
+      _sources,
+      check,
+      palantirjavaformatOptions(),
+      palantirjavaformatClasspath(),
+      palantirjavaformatJvmArgs()
+    )
+  }
+
+  /**
+   * Classpath for running Palantir Java Format.
+   */
+  def palantirjavaformatClasspath: T[Loose.Agg[PathRef]] = T {
+    defaultResolver().resolveDeps(
+      Agg(ivy"com.palantir.javaformat:palantir-java-format:${palantirjavaformatVersion()}")
+    )
+  }
+
+  /**
+   * JVM arguments for running Palantir Java Format. Defaults to values prescribed in
+   * "[[https://github.com/palantir/palantir-java-format/issues/548 Broken on Java 16]]".
+   */
+  def palantirjavaformatJvmArgs: T[Seq[String]] = T {
+    Seq(
+      "--add-exports",
+      "jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED",
+      "--add-exports",
+      "jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED",
+      "--add-exports",
+      "jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED",
+      "--add-exports",
+      "jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED",
+      "--add-exports",
+      "jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED"
+    )
+  }
+
+  /**
+   * Path to options file for Palantir Java Format CLI. Defaults to `millSourcePath` `/` `palantirjavaformat.options`.
+   */
+  def palantirjavaformatOptions: T[PathRef] = Task.Source(
+    millSourcePath / "palantirjavaformat.options"
+  )
+
+  /**
+   * Palantir Java Format version. Defaults to `2.50.0`.
+   */
+  def palantirjavaformatVersion: T[String] = T {
+    "2.50.0"
+  }
+}
+object PalantirJavaFormatModule extends ExternalModule with PalantirJavaFormatModule {
```

Comment:
Does this need to inherit from the full `PalantirJavaFormatModule`? `PalantirJavaFormatModule` pulls in `JavaModule`, and I don't think this external module needs to be able to compile java sources and package it etc... Maybe we can extract a more lightweight `trait` that contains the shared stuff for both `trait PalantirJavaFormatModule` and `object PalantirJavaFormatModule`, whereas only `trait PalantirJavaFormatModule` inherits from `JavaModule`

Also, we should define a `def defaultCommandName = "formatAll"`, so people can call this via `mill.contrib.palantirjavaformat.PalantirJavaFormatModule/` without needing to pass in the `formatAll` suffix explicitly